astro/src/content/quickstarts/quickstart-javascript-express-web.mdx

```diff
@@ -0,0 +1,320 @@
+
+<Aside type="note">
+In this file we will use the package [pkce-challenge](https://www.npmjs.com/package/pkce-challenge) to generate a Proof Key for Code Exchange (PKCE) challenge pair. Read more about [PKCE](https://www.oauth.com/oauth2-servers/pkce/authorization-request/).
+</Aside>
+
+`index.ts`
+<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
+  lang="ts"
+  tags="homepage"/>
+
+#### Login Route
+
+In the Changebank application there is a `Login` button, when clicked will open the `/login` route (you can find this in `complete-application/templates/home.html`). Have this redirect them to FusionAuth so that FusionAuth can present them with a login page.
+
+`index.ts`
+<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
+  lang="ts"
+  tags="login"/>
+
+You can always find this url if you view the details of your Application within FusionAuth, under "OAuth2 & OpenID Connect Integration details" you will find the property `OAuth IdP login URL`. The endpoint that you will hit on FusionAuth is `/oauth2/authorize` below you can find a list of the needed parameters for FusionAuth's [OAuth Endpoints](https://fusionauth.io/docs/v1/tech/oauth/endpoints)
+
+- client_id: Your `applicationId`
+- response_type: `code` since we are using the code flow
+- redirect_uri: Where your FusionAuth will redirect back to during the Oauth flow, in our example this is our route `oauth-redirect`
+- state: Created in our homepage route using some crazy math and stored in our cookie as `stateValue`
+- code_challenge: Created in our homepage route using `pkce-challenge` stored in our cookie as `challenge`, this is one half of the PKCE 
+
+#### OAuth Redirect
+
+When FusionAuth redirects back to our Express application it will return to our `/oauth-redirect` route. In this response it will have two query parameters that we need to further examine `code` and `state`. 
+
+1. Make sure to check that the `stateValue` that we stored in our cookie matches the `state` query parameters
+1. Request to the FusionAuth Token endpoint to exchange the authorization `code` returned from the Authorize endpoint and a `verifier` from our `userSession` cookie for an access token.
+1. Retrieves the user by passing the access token to FusionAuth. (This method does not use an API key, instead it uses a JSON Web Token (JWT) for authentication)
+1. Store access_token details in `userToken` cookie.
+1. Store retrieved user in `userDetails` cookie (note for our example this is not a secure cookie and JS can access).
+1. Redirect to `/account` route.
+
+`index.ts`
+<RemoteCode url={frontmatter.coderoot + "/complete-application/src/index.ts"}
```

Comment:
I think we should use [`timingSafeEqual()`](https://nodejs.org/api/crypto.html#cryptotimingsafeequala-b) instead of `if (stateFromFusionAuth !== userSessionCookie?.stateValue) {`

Reply:
This is a string comparison, how does this work for that?

Reply:
It's just a time-constant string comparison, so we could do something like this to avoid leaking information about the state string to an attacker. But maybe I'm being overzealous.

```ts
if (!timingSafeEqual(stateFromFusionAuth, userSessionCookie?.stateValue)) {
```

Reply:
This seems overkill to me as it will be on the server.